rust/rope/src/compare.rs

```diff
@@ -0,0 +1,550 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Fast comparison of rope regions, principally for diffing.
+
+#[cfg(target_arch = "x86")]
+use std::arch::x86::*;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::*;
+
+use rope::{BaseMetric, Rope, RopeInfo};
+use tree::{Cursor};
+
+// https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_cmpestrm
+const SSID_OPTS: i32 = _SIDD_UBYTE_OPS | _SIDD_CMP_EQUAL_EACH | _SIDD_UNIT_MASK | _SIDD_NEGATIVE_POLARITY;
+
+const SSE_STRIDE: usize = 16;
+
+/// Given two 16-byte slices, returns a bitmask where the 1 bits indicate
+/// the positions of non-equal bytes.
+///
+/// The least significant bit in the mask refers to the byte in position 0;
+/// that is, you read the mask right to left.
+///
+/// # Examples
+///
+/// ```
+/// use xi_rope::compare::fast_cmpestr_mask;
+/// let one = "aaaaaaaaaaaaaaaa";
+/// let two = "aa3aaaaa9aaaEaaa";
+/// let exp = "0001000100000100";
+/// let result = format!("{:016b}", fast_cmpestr_mask(one.as_bytes(), two.as_bytes()));
+/// assert_eq!(result.as_str(), exp);
+/// ```
+///
+#[inline(always)]
```

Comment:
This should _probably_ be protected by a `#[target_feature(enable = "sse4.2")]`. See [memchr-rs](https://github.com/BurntSushi/rust-memchr/blob/master/src/x86/sse42.rs#L24) source for guidance.

This function is also unsafe, because it is UB if the CPU doesn't support SSE4.2.